core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -3854,92 +3854,134 @@ object Stream extends StreamLowPriority {
-          def endWithResult(result: Either[Throwable, Unit]): F[Unit] =
-            result match {
-              case Right(()) => decrementRunning
-              case Left(err) => stop(Some(err)) >> decrementRunning
-            }
-
-          def sendToChannel(str: Stream[F, O]) = str.chunks.foreach(x => outputChan.send(x).void)
-
-          // runs one inner stream, each stream is forked.
-          // terminates when killSignal is true
-          // failures will be propagated through `done` Signal
-          // note that supplied scope's resources must be leased before the inner stream forks the execution to another thread
-          // and that it must be released once the inner stream terminates or fails.
           def runInner(inner: Stream[F, O], outerScope: Scope[F]): F[Unit] =
             F.uncancelable { _ =>
-              outerScope.lease.flatMap { lease =>
-                available.acquire >>
-                  incrementRunning >>
+              outerScope.lease
+                .flatTap(_ => available.acquire >> incrementRunning)
+                .flatMap { lease =>
                   F.start {
-                    // Note that the `interrupt` must be AFTER the send to the sync channel,
-                    // otherwise the process may hang to send last item while being interrupted
-                    val backInsertions = untilDone(sendToChannel(inner))
-                    for {
-                      r <- backInsertions.compile.drain.attempt
-                      cancelResult <- lease.cancel
-                      _ <- available.release
-                      _ <- endWithResult(CompositeFailure.fromResults(r, cancelResult))
-                    } yield ()
+                    inner.chunks
+                      .evalMap(s => output.send(s).void)
+                      .interruptWhen(done.map(_.nonEmpty))
+                      .compile
+                      .drain
+                      .guaranteeCase {
+                        case Outcome.Succeeded(fu) =>
+                          lease.cancel.flatMap {
+                            case Left(t)   => stop(Some(t))
+                            case Right(()) => outcomes.send(fu).void
+                          } >> available.release >> decrementRunning
+
+                        case Outcome.Errored(t) =>
+                          lease.cancel.flatMap { cancelResult =>
+                            (CompositeFailure.fromResults(Left(t), cancelResult) match {
+                              case Left(t)   => stop(Some(t))
+                              case Right(()) => F.unit
+                            })
+                          } >> available.release >> decrementRunning
+
+                        case Outcome.Canceled() =>
+                          lease.cancel.flatMap {
+                            case Left(t)   => stop(Some(t))
+                            case Right(()) => F.unit
+                          } >> available.release >> decrementRunning
```

Comment:
Definitely, feel free to suggest other places where this can be done. I was mainly focused on getting this to work first, I think it's safe to iterate on the code now.

Reply:
Done.